storybookjs/storybook
# CLI: Improve support of mono repositories

Closes https://github.com/storybookjs/storybook/issues/21216
Closes https://github.com/storybookjs/storybook/issues/21710



## What I did

Improved support for mono repositories to apply the "wrapForPnp" wrapper also in mono repository environments. Additionally, an auto migration was added.

Improved support for projects which are part of a workspace configured by yarn, npm, pnpm, lerna, turbo or rush

## How to test

- Go into a project which is part of a mono repository setup or where yarn pnp is enabled.
- If the project is a yarn workspace with pnp enabled, please run `yarn exec <storybook-project-folder>/code/lib/cli/bin/index.js automigrate` for the automigration or `yarn exec <storybook-project-folder>/code/lib/cli/bin/index.js init` to initialize Storybook. Otherwise, you can ditch the `yarn exec` part.
- Please check whether the values for the configuration fields in `main.ts` are properly wrapped by the `wrapForPnp` or `getAbsolutePath` wrapper.
- Play around with the following origin scenarios:
  - a wrapForPnp function is already in place, but is not used to e.g. wrap addon values
  - test it for `main.ts` and `main.js` files.

Tested scenarios:
- init
- auto-upgrade
  - where no fields were transformed
  - where all fields were already transformed (wrap-require automigration skipped)
  - where some fields were transformed (do not introduce a new wrapper if `getAbsolutePath` or `wrapForPnp` functions are in place)
  - checked for Typescript and Javascript

## Checklist



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers


`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`
